test/test_data/test_utils.py

```diff
@@ -0,0 +1,107 @@
+
+
+def test_validate_dataset_compression_arg():
+
+    data_compression_args = validate_dataset_compression_arg({}, 10)
+    # check whether the function uses default args
+    # to fill in case args is empty
+    assert data_compression_args is not None
+
+    # assert memory allocation is an integer after validation
+    assert isinstance(data_compression_args['memory_allocation'], int)
+
+    # check whether the function raises an error
+    # in case an unknown key is in args
+    with pytest.raises(ValueError, match=r'Unknown key in dataset_compression, .*'):
+        validate_dataset_compression_arg({'not_there': 1}, 1)
+
+    # check whether the function raises an error
+    # in case memory_allocation is not int or float is in args
+    with pytest.raises(ValueError, match=r"key 'memory_allocation' must be an `int` or `float`.*"):
+        validate_dataset_compression_arg({'memory_allocation': 'not int'}, 1)
+
+    # check whether the function raises an error
+    # in case memory_allocation is an int greater than memory limit
+    with pytest.raises(ValueError, match=r"key 'memory_allocation' if int must be in.*"):
+        validate_dataset_compression_arg({'memory_allocation': 1}, 0)
+
+    # check whether the function raises an error
+    # in case memory_allocation is a float greater than 1
+    with pytest.raises(ValueError, match=r"key 'memory_allocation' if float must be in.*"):
+        validate_dataset_compression_arg({'memory_allocation': 1.5}, 0)
+
+    # check whether the function raises an error
+    # in case an unknown method is passed in args
+    with pytest.raises(ValueError, match=r"key 'methods' can only contain .*"):
+        validate_dataset_compression_arg({'methods': 'unknown'}, 1)
+
+    # check whether the function raises an error
+    # in case an unknown key is in args
+    with pytest.raises(ValueError, match=r'Unknown type for `dataset_compression` .*'):
+        validate_dataset_compression_arg(1, 1)
+
+
+def test_error_raised_reduce_precision():
+    # check whether the function raises an error
+    # in case X is not an expected type
+    with pytest.raises(ValueError, match=r'Unrecognised data type of X, expected data type to .*'):
+        reduce_precision(X='not expected')
+
+
+def _verify_dataset_compression_mapping(mapping):
+    assert isinstance(mapping, Mapping)
+    assert 'methods' in mapping
+    assert 'memory_allocation' in mapping
+
+
+@pytest.mark.parametrize('memory_limit', [2048])
+def test_get_dataset_compression_mapping(memory_limit):
+    """
+    Tests the functionalities of `get_dataset_compression_mapping`
+    """
+    dataset_compression_mapping = get_dataset_compression_mapping(
+        dataset_compression=True,
+        memory_limit=memory_limit)
+    _verify_dataset_compression_mapping(dataset_compression_mapping)
+
+    dataset_compression_mapping = get_dataset_compression_mapping(
+        dataset_compression={'memory_allocation': 0.01, 'methods': ['precision']},
```

Comment:
Can we test whether the mapping really follows this rule?

Reply:
sure